### PR TITLE
Optimize some areas of 2D rendering (3.2)

### PR DESCRIPTION
**Pull request by RandomShaper**
Commits in this PR:

#### Use direct rendering when joining is disabled
The batching renderer allows you to disable joining. That's useful if your game doesn't benefit from joining for some reason. But you'd run anyway the penalty of the two-stage process (recording and then rendering recorded single item chunks). This commit adds a fast path for the batching render pipeline (to also get other benefits from it, like light scissoring, and because the legacy one will be removed at some point).

#### Improve light scissoring
This commit does two things:
1. Tracks if GL scissoring is on or off in the context of 2D lighting, to save some calls to the driver.
1. Rounds up the size of the scissor rect to avoid 1-pixel seams at bottom and/or right.

#### Optimize light intersection tests
  Normally, intersection tests used in 2D lighting need first to transform the axis-aligned light's bounding rect to world coordinates. This was formerly being done per every intersection test. This commit introduces a new variant of the intersection test method that allows the caller to pass an array of the points already transformed. This is then leveraged in the rasterizer to benefit both GL2/GL3 regular light intersection tests and specifically the GL2 batching renderer when it comes to light scissoring testing (you can afford now a lower threshold since computations are less expensive).

### CODE REVIEW

File: core/math/rect2.cpp [hunk 1]
```diff
@@ -95,26 +95,31 @@ bool Rect2::intersects_segment(const Point2 &p_from, const Point2 &p_to, Point2
 
 bool Rect2::intersects_transformed(const Transform2D &p_xform, const Rect2 &p_rect) const {
 
-	//SAT intersection between local and transformed rect2
-
-	Vector2 xf_points[4] = {
+	Vector2 p_xf_points[4] = {
```
Comment (groud) [inline core/math/rect2.cpp:98]: Well, usually we only prepend parameters with `p_`. That's kind of confusing if you prepend a variable too.

**Comment by lawnjelly** [comment]
> Use direct rendering when joining is disabled

Could there be a misunderstanding here - when item joining is off, the batching still does command joining, which is how the majority of batching gains occur (i.e. text / tilemaps etc). Item joining is more a secondary thing to help with things like sprites (2-3x speedup), and will never give the gains that command batching gives (10x +).

We could apply light scissoring to the legacy path on the other hand, if that was desired. I don't think I thought it was worth it at the time, in order to keep the legacy path as a 'ground truth', and also because most people would be using batching.

> Improve light scissoring

This could be worth doing, especially the adding an extra bit to width and height in case (I haven't checked this through yet). Someone did mention something about an artifact from scissoring I think on discord, but as yet no issue about it, so that could have been related.

> Optimize light intersection tests

This sounds sensible.

Overall though it might be an idea to hold off on this temporarily until GLES3 batching is in. The reason being the whole structure of the batching changes significantly, and moves to a template (it actually becomes a bit easier to follow if I remember right). So it would be easier to fit these in after that change than before.

Although perhaps some of the light intersection test ideas could be changed in the visual server in the meantime if that didn't affect the renderers?

Also in terms of the roadmap it sounds as though reduz is keen to try writing batching in a different way for GLES3 in 4.x, although much of the current batching will probably be ported for GLES2 in 4.x, with possible significant changes though depending on whether we drop support for custom shaders.

File: core/math/rect2.cpp (hunk 1 above)
Comment (RandomShaper) [reply core/math/rect2.cpp]: Oops. A case of uncritical copy-paste. Fixed!

Comment (groud) [reply core/math/rect2.cpp]: That's what I suspected. ^^

**Comment by RandomShaper** [comment]
> > Use direct rendering when joining is disabled
> 
> Could there be a misunderstanding here - when item joining is off, the batching still does command joining, which is how the majority of batching gains occur (i.e. text / tilemaps etc). Item joining is more a secondary thing to help with things like sprites (2-3x speedup), and will never give the gains that command batching gives (10x +).
> 
> We could apply light scissoring to the legacy path on the other hand, if that was desired. I don't think I thought it was worth it at the time, in order to keep the legacy path as a 'ground truth', and also because most people would be using batching.
> 

I think my wording was not precise. I took the "joining disabled" part from the description of the setting:
image

Adding to the legacy path features that are now unique to the batching renderer may make sense, but isn't it expected to get removed eventually?

All that said, the idea in this commit still has a point, does it? It avoids going through the two-stage process when item joining is disabled. It's true that it should not be called "direct rendering" since batching still happens.

> > Optimize light intersection tests
> 
> This sounds sensible.
> 
> Overall though it might be an idea to hold off on this temporarily until GLES3 batching is in. The reason being the whole structure of the batching changes significantly, and moves to a template (it actually becomes a bit easier to follow if I remember right). So it would be easier to fit these in after that change than before.
> 
> Although perhaps some of the light intersection test ideas could be changed in the visual server in the meantime if that didn't affect the renderers?

I think so. The improved intersection tests only involve a marginal change to the batching renderer.

> Also in terms of the roadmap it sounds as though reduz is keen to try writing batching in a different way for GLES3 in 4.x, although much of the current batching will probably be ported for GLES2 in 4.x, with possible significant changes though depending on whether we drop support for custom shaders.

When I know the answer to these questions, I'll propose a plan to create a new PR with the changes that can be merged now and hold the others off to re-apply them to the new rendering code once it's ready.

**Comment by lawnjelly** [comment]
Ah sorry I get you, you want to avoid recording items if they aren't being joined.

This may be worth doing yes, although the trade off will have to be examined whether it in practice provides any non-negligible speed improvement, versus having less code paths. But either way I think we should hold off on that until we have the template in. :+1:

File: drivers/gles2/rasterizer_canvas_gles2.h [hunk 2]
```diff
@@ -129,7 +129,12 @@ class RasterizerCanvasGLES2 : public RasterizerCanvasBaseGLES2 {
...
 	// batch item may represent 1 or more items
 	struct BItemJoined {
-		uint32_t first_item_ref;
+		union {
+			uint32_t first_item_ref;
+			Item *single_item;
+		};
+		// >= 1: standard case, first_item_ref is valid
+		// 0: joining disabled, single_item is valid (this BItemJoined is used as an adapter to a single item)
 		uint32_t num_item_refs;
```
Comment (lawnjelly) [inline drivers/gles2/rasterizer_canvas_gles2.h:138]: Note that Item * will be 64 bit on 64 bit platforms, which will make the structure larger and affect alignment.

**Comment by lawnjelly** [comment]
I think also when we have a look at these later, we really need to get some performance figures with and without each change, in different scenarios. Any optimization effort has to be guided by performance testing.

All this code is very performance sensitive, and I'm extremely aware that improving things in one scenario can lead to a drop of performance in others. There is a big danger of optimizing for a particular game, at the expense of the general case.

Also bear in mind in the case of adding code complexity, there would normally have to be a significant improvement to counterbalance added complexity in the code. The batching code is already pretty convoluted, and maintenance / ease of support / dealing with regressions is a big concern - the less codepaths / special cases the better.

Personally speaking I would find it easier rather than trying to combine several (independent) changes in one PR, if we had one PR per change then they could be evaluated independently. Some of the changes look as though they may be quite worthwhile (for instance the pixel border on the scissor, if there is a minimum reproduction project showing this to be a problem, so we can examine it in more detail), whereas others may turn out to be less useful, or more testing may be needed.

For instance adding scissor border is no problem to merge (if it did fix a bug), but state optimizing the glScissor state could lead to regressions if something outside the immediate area of code changes the scissor state or relied on it being in a particular state, so would have to be tested more rigorously (this is why it wasn't done already btw, however that is not a reason why we shouldn't get it working :) ). Again some performance figures would be helpful to know how much it helps.

Also about the removal of the legacy path .. given how close we are to moving on to work on 4.x, and how this will be significantly rewritten in 4.x GLES2 and completely different in 4.x GLES3, I'm no longer convinced removal of the legacy path is a pressing issue. At this stage there is little cost to keeping it, and it has proved useful for support issues (finding regressions with old behaviour) and may help the occasional game that does not benefit at all from batching (e.g. using incompatible custom shaders throughout).

Also note that the kind of percentage improvements possible in these areas are likely to be dwarfed by improvements made by increasing the applicability of the batching (e.g. applying it to more custom shaders, more primitives etc), which can be done once GLES3 has been sorted.

**Comment by RandomShaper** [comment]
Thanks for your detailed explanation.

I did these for my game and though they could be useful for general purpose, but that's not the case. What I'll do it close this PR.

- The + 0.5 fix for the light scissoring is still a good thing, but I guess you can just drop it into your new code.
- If I have time some day to benchmark the transform optimization, I'll revive it myself in a new PR.

**Pull request closed without merging**